Add verbose-recursive flag to kubectl explain to print all the fields

ah8ad3 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Prints out all the details in explain so user can work around it, in this issue for example user can use `grep` to find a specific keyword inside the resource.
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubectl/issues/1557


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add new template to kubectl explain to print fields recursive with all the details with new graph styling.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
kubectl explain pods.spec.containers.startupProbe --recursive=false
```

```
KIND:       Pod
VERSION:    v1

FIELD: startupProbe <Probe>

DESCRIPTION:
    StartupProbe indicates that the Pod has successfully initialized. If
    specified, no other probes are executed until this completes successfully.
    If this probe fails, the Pod will be restarted, just as if the livenessProbe
    failed. This can be used to provide different probe parameters at the
    beginning of a Pod's lifecycle, when it might take a long time to load data
    or warm a cache, than during steady-state operation. This cannot be updated.
    More info:
    https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle#container-probes
    Probe describes a health check to be performed against a container to
    determine whether it is alive or ready to receive traffic.
    
FIELDS:
* exec  <ExecAction>
| Exec specifies the action to take.
| 
| * command     <[]string>
| | Command is the command line to execute inside the container, the working
| | directory for the command  is root ('/') in the container's filesystem. The
| | command is simply exec'd, it is not run inside a shell, so traditional shell
| | instructions ('|', etc) won't work. To use a shell, you need to explicitly
| | call out to that shell. Exit status of 0 is treated as live/healthy and
| | non-zero is unhealthy.
| | 
* failureThreshold      <integer>
| Minimum consecutive failures for the probe to be considered failed after
| having succeeded. Defaults to 3. Minimum value is 1.
| 
* grpc  <GRPCAction>
| GRPC specifies an action involving a GRPC port.
| 
| * port        <integer> -required-
| | Port number of the gRPC service. Number must be in the range 1 to 65535.
| | 
| * service     <string>
| | Service is the name of the service to place in the gRPC HealthCheckRequest
| | (see https://github.com/grpc/grpc/blob/master/doc/health-checking.md).
| | 
| | If this is not specified, the default behavior is defined by gRPC.
| | 
* httpGet       <HTTPGetAction>
| HTTPGet specifies the http request to perform.
| 
| * host        <string>
| | Host name to connect to, defaults to the pod IP. You probably want to set
| | "Host" in httpHeaders instead.
| | 
| * httpHeaders <[]HTTPHeader>
| | Custom headers to set in the request. HTTP allows repeated headers.
| | 
| | * name      <string> -required-
| | | The header field name. This will be canonicalized upon output, so
| | | case-variant names will be understood as the same header.
| | | 
| | * value     <string> -required-
| | | The header field value
| | | 
| * path        <string>
| | Path to access on the HTTP server.
| | 
| * port        <IntOrString> -required-
| | Name or number of the port to access on the container. Number must be in the
| | range 1 to 65535. Name must be an IANA_SVC_NAME.
| | 
| * scheme      <string>
| | enum: HTTP, HTTPS
| | Scheme to use for connecting to the host. Defaults to HTTP.
| | 
| | Possible enum values:
| |  - `"HTTP"` means that the scheme used will be http://
| |  - `"HTTPS"` means that the scheme used will be https://
| | 
* initialDelaySeconds   <integer>
| Number of seconds after the container has started before liveness probes are
| initiated. More info:
| https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle#container-probes
| 
* periodSeconds <integer>
| How often (in seconds) to perform the probe. Default to 10 seconds. Minimum
| value is 1.
| 
* successThreshold      <integer>
| Minimum consecutive successes for the probe to be considered successful after
| having failed. Defaults to 1. Must be 1 for liveness and startup. Minimum
| value is 1.
| 
* tcpSocket     <TCPSocketAction>
| TCPSocket specifies an action involving a TCP port.
| 
| * host        <string>
| | Optional: Host name to connect to, defaults to the pod IP.
| | 
| * port        <IntOrString> -required-
| | Number or name of the port to access on the container. Number must be in the
| | range 1 to 65535. Name must be an IANA_SVC_NAME.
| | 
* terminationGracePeriodSeconds <integer>
| Optional duration in seconds the pod needs to terminate gracefully upon probe
| failure. The grace period is the duration in seconds after the processes
| running in the pod are sent a termination signal and the time when the
| processes are forcibly halted with a kill signal. Set this value longer than
| the expected cleanup time for your process. If this value is nil, the pod's
| terminationGracePeriodSeconds will be used. Otherwise, this value overrides
| the value provided by the pod spec. Value must be non-negative integer. The
| value zero indicates stop immediately via the kill signal (no opportunity to
| shut down). This is a beta field and requires enabling
| ProbeTerminationGracePeriod feature gate. Minimum value is 1.
| spec.terminationGracePeriodSeconds is used if unset.
| 
* timeoutSeconds        <integer>
| Number of seconds after which the probe times out. Defaults to 1 second.
| Minimum value is 1. More info:
| https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle#container-probes
```

ah8ad3 committed:
Add verbose-recursive flag to kubectl explain to print all the fields
with description.

apelisse commented on staging/src/k8s.io/kubectl/pkg/explain/v2/funcs.go:
```diff
@@ -236,5 +236,19 @@ func WithBuiltinTemplateFuncs(tmpl *template.Template) *template.Template {
 		},
+		// styleVerboseRecursive creates an string for styling the verboseRecursive based
+		// on if the line is a filed or description/empty line.
+		"styleVerboseRecursive": func(level int, field bool) string {
```
I'm curious what Alex thinks, but maybe we can embed that in the `indent` function? along with the flag so that we don't have to pass that to the schema? Curious about your opinions.

apelisse replied:
Or provide a way to indent in the template in a more flexible way?

ah8ad3 replied:
The reason i've created a func here was that we are going to use some brand new styling and maybe we don't need to change the `indent` or template so much and use it like this. But if you or Alex thinks merging it to template or `indent` function would be better approach i'm happy to change it.

ah8ad3 commented on staging/src/k8s.io/kubectl/pkg/explain/v2/explain.go:
```diff
@@ -55,6 +56,7 @@ func printModelDescriptionWithGenerator(
 	client openapi.Client,
 	gvr schema.GroupVersionResource,
 	recursive bool,
+	verboseRecursive bool,
```
You are completely right about this, TBH at first i wasn't sure about it too, asked about it in the original issue and didn't get any answer so i decide to put some thing here to discuss about it.

As you mention these two flags can not be used with each other, i think changing the type of `recursive` from `bool` to `string` and accepting things like `short`, `detailed` would make better sense to me. 
Like 
`kubectl explain pods.spec.containers.startupProbe --recursive=detailed` 
or 
`kubectl explain pods.spec.containers.startupProbe --recursive=short`

But it will change the usage of `--recursive=true` and i'm not sure how we deal with that. If it's ok i can implement that.
But also if we want to use depth here too idk how to catch it with one flag? maybe using something like this 
`--recursive=detailed=2` the same approach the feature gate is using 
`--feature-gates=...,GracefulNodeShutdown=true` but i'm not sure if end user will be happy about this or not.

ah8ad3 commented on staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go:
```diff
@@ -51,6 +51,9 @@ var (
 		kubectl explain pods --recursive
 
+		# Get all the fields in the resource with their descriptions
+		kubectl explain pods --versobe-recursive
```
Mentioned the depth in this comment https://github.com/kubernetes/kubernetes/pull/124218#discussion_r1556216753

alexzielenski replied:
So this is being added because we already have 2 implicit modes:

1. Long-form
2. Recursive

However we infer the value of the long form view based on the value of recursive. This PR seems to aim to separate the concerns. I think for this case verbose may be overloaded to cause confusion with extended logging output.  Maybe another flag is justified, so we could have 4 modes:

1. Recursive Long
2. Recursive Short
3. Shallow Long
4. Shallow Short

To keep behavior the same however, we'd need to adopt the uncommon/possibly confusing pattern of defaulting the value of a new long-form flag based on the value of the recursive flag. I'm not sure how to best mitigate that

apelisse replied:
Should this be a seperate template to be honest?

alexzielenski replied:
> Should this be a seperate template to be honest?

I think not, the template already has a concept of long vs short display along with recursion but the separation doesn't bubble to the interface. Adding this doesnt introduce much complexity IMO

alexzielenski replied:
>  I don't think it has a lot of value does it?

Unfortunately we dont have a lot of data on who is using it and in the absense of that Id be uncomfortable taking this away. Some might find it convenient and easy to read to list out all the fields for a quick reminder 🤷🏽

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go:
```diff
@@ -67,9 +70,10 @@ var (
-	Recursive  bool
+	CmdParent        string
+	APIVersion       string
+	Recursive        bool
+	VerboseRecursive bool
```
I'd prefer if possible for this to be refactored so that the new flag introduces an orthogonal concept. Right now there is overlap with `recursive`. For example for verbose recursive we might want a `Long-form` flag to be set and the `Recursive` flag.

ah8ad3 replied:
Thanks @alexzielenski , if i understand it well we should use something like this `kubectl explain pods --recursive --long-form`. I am not sure what should i print when user does this: `kubectl explain pods --long-form=false`.

alexzielenski commented on staging/src/k8s.io/kubectl/pkg/explain/v2/templates/plaintext.tmpl:
```diff
@@ -196,18 +198,37 @@ Takes dictionary as argument with keys:
     schema: openapiv3 JSON schema which contains the field
     name: name of field
     short: limit printing to a single-line summary
+    verboseRecursive: style the line and print everything
```
Similarly to my other comment the refactoring would extend to the template where we remove references to verbose recursive. Verbose recursive mode would set `short` here to false and possibly another parameter to control the indentation styling

ah8ad3 replied:
Thanks @alexzielenski i'll for sure change that.

ah8ad3 committed:
Add plaintext_short template to separate long vs short view of resources in kubectl explain.
Short means all the fields without any detail.
Long means all the field with all the details.

For now we can change the template with --recursive which means:
--recursive=true goes to the short version with recursive
--recursive=false goes to the long version with recursive
so this way API surface stays the same.

refactor: remove unwanted comments

refactor: remove unwanted comments

ah8ad3 commented:
Comments addressed in last commit, PTAL thanks.

ah8ad3 commented on staging/src/k8s.io/kubectl/pkg/cmd/explain/explain.go:
```diff
@@ -188,6 +195,15 @@ func (o *ExplainOptions) Run() error {
+		// Use recursive flag to decide which template to render
+		if o.Flags.Changed("recursive") && !o.Recursive {
+			// Use default long format template
+			o.Recursive = true
```
My concern is this part, maybe users cant understand naming.

ah8ad3 committed:
change signature of indent func to make it more general purpose

ah8ad3 commented on staging/src/k8s.io/kubectl/pkg/explain/v2/funcs.go:
```diff
@@ -123,8 +123,21 @@ func WithBuiltinTemplateFuncs(tmpl *template.Template) *template.Template {
 			}
 		},
-		"indent": func(amount int, str string) string {
+		"indent": func(amount int, isGraph bool, isField bool, str string) string {
```
Thanks that's much better, updated.

ah8ad3 commented:
/ok-to-test
